moped-editor/src/views/projects/projectView/MilestoneTemplateModal.js

```diff
@@ -0,0 +1,198 @@
+  TextField,
+} from "@material-ui/core";
+import { Autocomplete } from "@material-ui/lab";
+import { makeStyles } from "@material-ui/core";
+import CloseIcon from "@material-ui/icons/Close";
+import AddCircle from "@material-ui/icons/AddCircle";
+import { returnSignalPHBMilestoneTemplate } from "../../../utils/timelineTemplates";
+
+import { ADD_PROJECT_MILESTONE } from "../../../queries/project";
+import { useMutation } from "@apollo/client";
+
+const useStyles = makeStyles((theme) => ({
+  dialogTitle: {
+    color: theme.palette.primary.main,
+    fontFamily: theme.typography.fontFamily,
+    display: "flex",
+    justifyContent: "space-between",
+    alignItems: "center",
+  },
+}));
+
+const templateChoices = ["AMD", "PDD"];
+
+/**
+ * useMemo hook to choose milestone options
+ * @return {Object[]}
+ */
+const useMilestoneOptions = (template, projectId, selectedMilestonesIds) =>
+  useMemo(() => {
+    if (template === "AMD") {
+      return returnSignalPHBMilestoneTemplate(projectId);
+      // } else if (template === "PDD") {
+      //   // etc return [];
+    } else {
+      return [];
+    }
+  }, [template, projectId]);
+
+/**
+ * useMemo hook to filter out already selected milestones
+ * @return {Object[]}
+ */
+const useMilestoneSelections = (milestonesList, selectedMilestonesIds) =>
+  useMemo(() => {
+    return milestonesList.filter(
+      (option) => !selectedMilestonesIds.includes(option.milestone_id)
+    );
+  }, [milestonesList, selectedMilestonesIds]);
+
+const MilestoneTemplateModal = ({
+  isDialogOpen,
+  handleDialogClose,
+  projectId,
+  milestoneNameLookup,
+  selectedMilestones,
+  refetch,
+}) => {
+  const classes = useStyles();
+
+  const [template, setTemplate] = useState(null);
+  const [milestonesToAdd, setMilestonesToAdd] = useState([]);
+
+  const [addProjectMilestone] = useMutation(ADD_PROJECT_MILESTONE);
+
+  // Array of milestone ids already in moped_proj_milestones
+  const selectedMilestonesIds = selectedMilestones.map(
+    (milestone) => milestone.milestone_id
+  );
+
+  const milestonesList = useMilestoneOptions(template, projectId);
+
+  const filteredMilestonesList = useMilestoneSelections(
+    milestonesList,
+    selectedMilestonesIds
+  );
+
+  useEffect(() => {
+    setMilestonesToAdd([...milestonesList]);
+  }, [milestonesList]);
+
+  // checks if milestone is in list of milestones to add
+  // if in list, remove. if not, add.
+  const handleToggle = (milestone) => {
+    const currentIndex = milestonesToAdd.indexOf(milestone);
+    const newChecked = [...milestonesToAdd];
+
+    if (currentIndex === -1) {
+      newChecked.push(milestone);
+    } else {
+      newChecked.splice(currentIndex, 1);
+    }
+
+    setMilestonesToAdd(newChecked);
+  };
+
+  // calls function to close dialog and also resets state
+  const closeDialog = () => {
+    handleDialogClose();
+    setMilestonesToAdd([]);
+    setTemplate(null);
+  };
+
+  const handleAddMilestones = () => {
+    addProjectMilestone({
```

Comment:
Forgot to add this comment in the review, but, if you log `milestonesToAdd` here, you can see that duplicate milestones are queued up to be added to the project.

To reproduce:
1. Create a new project
2. Add all milestones using the AMD template except for one of them
3. Add the last remaining milestone using the AMD template again and see WSOD
4. The logged value of `milestonesToAdd` shows more than the one remaining option
5. The project has duplicates of previously added milestones

Reply:
John found a place in the code where I had a bug, I just pushed an update. I will run through these steps and hope that the update fixes this bug. thank you for a thorough listing of steps!

Reply:
Cool - just saw that and I'm taking another look so I can approve 🎉